File: C/Day-12/problem-1/problem-1-solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#define MAX_ROWS 1000
#define MAX_COLS 1000

// Directions for neighbors (right, down, left, up)
int directions[4][2] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

// Grid dimensions
int rows, cols;
char grid[MAX_ROWS][MAX_COLS];
bool visited[MAX_ROWS][MAX_COLS];

// Utility function to check bounds
bool is_within_bounds(int x, int y) {
    return x >= 0 && x < rows && y >= 0 && y < cols;
}

// Flood-fill to calculate area and perimeter
void flood_fill(int x, int y, char plant, int *area, int *perimeter) {
    visited[x][y] = true;
    (*area)++;

    for (int d = 0; d < 4; d++) {
        int nx = x + directions[d][0];
        int ny = y + directions[d][1];

        if (!is_within_bounds(nx, ny) || grid[nx][ny] != plant) {
            (*perimeter)++;
        } else if (!visited[nx][ny]) {
            flood_fill(nx, ny, plant, area, perimeter);
        }
    }
}

// Main function for Part 1
int main() {
    FILE *file = fopen("inputdata.txt", "r");
    if (!file) {
        perror("Error opening file");
        return EXIT_FAILURE;
    }

    rows = 0;
    while (fgets(grid[rows], MAX_COLS, file)) {
        grid[rows][strcspn(grid[rows], "\n")] = '\0';
        rows++;
    }
    fclose(file);

    cols = strlen(grid[0]);
    memset(visited, 0, sizeof(visited));

    int total_cost = 0;

    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            if (!visited[i][j]) {
                int area = 0, perimeter = 0;
                flood_fill(i, j, grid[i][j], &area, &perimeter);
                total_cost += area * perimeter;
            }
        }
    }

    printf("Part 1 Total Cost: %d\n", total_cost);
    return 0;
}

File: Go/Day-12/problem-1/problem-1-solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

// Region struct holds information about area and perimeter of a region
type Region struct {
	Area      int
	Perimeter int
}

// Helper function to check if a cell is within bounds and matches the region's character
func isValid(x, y int, grid [][]rune, visited [][]bool, char rune) bool {
	return x >= 0 && x < len(grid) && y >= 0 && y < len(grid[0]) && !visited[x][y] && grid[x][y] == char
}

// Perform a flood fill to calculate area and perimeter of a region
func floodFill(x, y int, grid [][]rune, visited [][]bool, char rune) Region {
	directions := [][]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}} // Up, Down, Left, Right
	queue := [][]int{{x, y}}
	visited[x][y] = true
	area, perimeter := 0, 0

	for len(queue) > 0 {
		cx, cy := queue[0][0], queue[0][1]
		queue = queue[1:]
		area++

		// Check all 4 directions
		for _, dir := range directions {
			nx, ny := cx+dir[0], cy+dir[1]
			if isValid(nx, ny, grid, visited, char) {
				visited[nx][ny] = true
				queue = append(queue, []int{nx, ny})
			} else if nx < 0 || ny < 0 || nx >= len(grid) || ny >= len(grid[0]) || grid[nx][ny] != char {
				// Increment perimeter if the neighboring cell is outside or a different type
				perimeter++
			}
		}
	}

	return Region{Area: area, Perimeter: perimeter}
}

func calculateCost(grid [][]rune) int {
	rows := len(grid)
	cols := len(grid[0])
	visited := make([][]bool, rows)
	for i := range visited {
		visited[i] = make([]bool, cols)
	}

	totalCost := 0
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			if !visited[i][j] {
				region := floodFill(i, j, grid, visited, grid[i][j])
				cost := region.Area * region.Perimeter
				totalCost += cost
			}
		}
	}

	return totalCost
}

func main() {
	file, err := os.Open("inputdata.txt")
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}
	defer file.Close()

	var grid [][]rune
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		grid = append(grid, []rune(line))
	}

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	totalCost := calculateCost(grid)
	fmt.Println("Total cost of fencing all regions:", totalCost)
}
